pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/JavaNode.java

```diff
@@ -42,4 +42,19 @@
     @Deprecated
     void setScope(Scope scope);
 
+    /**
+     * Returns the node representing the type declaration this node is
+     * found in. The type of that node is the type of the {@code this}
+     * expression.
+     *
+     * <p>This returns null for nodes that aren't enclosed in a type declaration.
+     * This includes {@linkplain ASTPackageDeclaration PackageDeclaration},
+     * This includes {@linkplain ASTImportDeclaration ImportDeclaration},
+     * {@linkplain ASTModuleDeclaration ModuleDeclaration},
+     * {@linkplain ASTCompilationUnit CompilationUnit}, and top-level
+     * {@linkplain ASTAnyTypeDeclaration AnyTypeDeclaration}s.
+     *
+     * @return the node of the enclosing type or <code>null</code>.
+     */
+    ASTAnyTypeDeclaration getEnclosingType();
```

Comment:
So obviously, if we're adding methods to interfaces on master, then we can do it for all interfaces which are not supposed to be extended by clients (AST). Meaning that for example, we can already pull some methods up from ASTMethodDeclaration onto ASTMethodOrConstructorDeclaration. I think this is sensible and would allow us to ease the transition to 7.0.0 for this API.